File: android/app/src/main/java/com/cochat/app/data/remote/SocketManager.kt
package com.cochat.app.data.remote

import com.cochat.app.data.model.AppNotification
import com.cochat.app.data.model.ChatSummary
import com.cochat.app.data.model.GroupInfo
import com.cochat.app.data.model.Message
import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.serialization.json.Json
import org.json.JSONObject

sealed class SocketEvent {
    data class PresenceUpdate(val userId: String, val status: String) : SocketEvent()
    data class MessageNew(val message: Message, val chatId: String) : SocketEvent()
    data class ChatUpdated(val chat: ChatSummary) : SocketEvent()
    data class ChatRead(val chatId: String, val userId: String) : SocketEvent()
    data class GroupUpdated(val chat: GroupInfo, val action: String) : SocketEvent()
    data class TypingUpdate(val chatId: String, val userId: String, val isTyping: Boolean) : SocketEvent()
    data class NotificationNew(val notification: AppNotification) : SocketEvent()
}

class SocketManager {
    private var socket: Socket? = null
    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    private val _events = MutableSharedFlow<SocketEvent>(0, 64)
    val events: SharedFlow<SocketEvent> = _events.asSharedFlow()

    fun connect(token: String) {
        if (socket?.connected() == true) return
        val options = IO.Options.builder().setAuth(mapOf("token" to token)).build()
        val s = IO.socket(java.net.URI.create(NetworkConfig.SOCKET_URL), options)

        s.on("presence:update") { args ->
            val o = args[0] as JSONObject
            emit(SocketEvent.PresenceUpdate(o.getString("userId"), o.getString("status")))
        }
        s.on("message:new") { args ->
            val o = args[0] as JSONObject
            val message = json.decodeFromString<Message>(o.getJSONObject("message").toString())
            emit(SocketEvent.MessageNew(message, o.getString("chatId")))
        }
        s.on("chat:updated") { args ->
            val o = args[0] as JSONObject
            val chat = json.decodeFromString<ChatSummary>(o.getJSONObject("chat").toString())
            emit(SocketEvent.ChatUpdated(chat))
        }
        s.on("chat:read") { args ->
            val o = args[0] as JSONObject
            emit(SocketEvent.ChatRead(o.getString("chatId"), o.getString("userId")))
        }
        s.on("group:updated") { args ->
            val o = args[0] as JSONObject
            val chat = json.decodeFromString<GroupInfo>(o.getJSONObject("chat").toString())
            emit(SocketEvent.GroupUpdated(chat, o.getString("action")))
        }
        s.on("typing:update") { args ->
            val o = args[0] as JSONObject
            emit(SocketEvent.TypingUpdate(o.getString("chatId"), o.getString("userId"), o.getBoolean("isTyping")))
        }
        s.on("notification:new") { args ->
            val o = args[0] as JSONObject
            val notification = json.decodeFromString<AppNotification>(o.getJSONObject("notification").toString())
            emit(SocketEvent.NotificationNew(notification))
        }

        s.connect()
        socket = s
    }

    fun emitTypingStart(chatId: String) = socket?.emit("typing:start", JSONObject().put("chatId", chatId))
    fun emitTypingStop(chatId: String) = socket?.emit("typing:stop", JSONObject().put("chatId", chatId))
    fun emitActivity() = socket?.emit("presence:activity")

    fun disconnect() {
        socket?.disconnect()
        socket?.off()
        socket = null
    }

    private fun emit(event: SocketEvent) {
        _events.tryEmit(event)
    }
}

File: ios/CoChat/Networking/SocketManager.swift
import Foundation
import SocketIO

enum AppSocketEvent {
    case presenceUpdate(userId: String, status: String)
    case messageNew(message: Message, chatId: String)
    case chatUpdated(chat: ChatSummary)
    case chatRead(chatId: String, userId: String)
    case groupUpdated(chat: GroupInfo, action: String)
    case typingUpdate(chatId: String, userId: String, isTyping: Bool)
    case notificationNew(notification: AppNotification)
}

@MainActor
final class AppSocketManager {
    private var manager: SocketManager?
    private var socket: SocketIOClient?
    private let decoder = JSONDecoder()

    /// Fan-out target; RealtimeStore assigns itself here in `start()`.
    var onEvent: ((AppSocketEvent) -> Void)?

    func connect(token: String) {
        guard socket?.status != .connected else { return }
        let manager = SocketManager(socketURL: NetworkConfig.socketURL, config: [.connectParams(["token": token]), .compress])
        self.manager = manager
        let socket = manager.defaultSocket
        self.socket = socket

        socket.on("presence:update") { [weak self] data, _ in
            guard let dict = data.first as? [String: Any],
                  let userId = dict["userId"] as? String,
                  let status = dict["status"] as? String else { return }
            self?.onEvent?(.presenceUpdate(userId: userId, status: status))
        }
        socket.on("message:new") { [weak self] data, _ in
            guard let self, let dict = data.first as? [String: Any],
                  let chatId = dict["chatId"] as? String,
                  let messageDict = dict["message"],
                  let message = self.decode(Message.self, from: messageDict) else { return }
            self.onEvent?(.messageNew(message: message, chatId: chatId))
        }
        socket.on("chat:updated") { [weak self] data, _ in
            guard let self, let dict = data.first as? [String: Any],
                  let chatDict = dict["chat"],
                  let chat = self.decode(ChatSummary.self, from: chatDict) else { return }
            self.onEvent?(.chatUpdated(chat: chat))
        }
        socket.on("chat:read") { [weak self] data, _ in
            guard let dict = data.first as? [String: Any],
                  let chatId = dict["chatId"] as? String,
                  let userId = dict["userId"] as? String else { return }
            self?.onEvent?(.chatRead(chatId: chatId, userId: userId))
        }
        socket.on("group:updated") { [weak self] data, _ in
            guard let self, let dict = data.first as? [String: Any],
                  let action = dict["action"] as? String,
                  let chatDict = dict["chat"],
                  let group = self.decode(GroupInfo.self, from: chatDict) else { return }
            self.onEvent?(.groupUpdated(chat: group, action: action))
        }
        socket.on("typing:update") { [weak self] data, _ in
            guard let dict = data.first as? [String: Any],
                  let chatId = dict["chatId"] as? String,
                  let userId = dict["userId"] as? String,
                  let isTyping = dict["isTyping"] as? Bool else { return }
            self?.onEvent?(.typingUpdate(chatId: chatId, userId: userId, isTyping: isTyping))
        }
        socket.on("notification:new") { [weak self] data, _ in
            guard let self, let dict = data.first as? [String: Any],
                  let notificationDict = dict["notification"],
                  let notification = self.decode(AppNotification.self, from: notificationDict) else { return }
            self.onEvent?(.notificationNew(notification: notification))
        }

        socket.connect()
    }

    func emitTypingStart(chatId: String) { socket?.emit("typing:start", ["chatId": chatId]) }
    func emitTypingStop(chatId: String) { socket?.emit("typing:stop", ["chatId": chatId]) }
    func emitActivity() { socket?.emit("presence:activity") }

    func disconnect() {
        socket?.disconnect()
        socket = nil
        manager = nil
    }

    private func decode<T: Decodable>(_ type: T.Type, from any: Any) -> T? {
        guard let data = try? JSONSerialization.data(withJSONObject: any) else { return nil }
        return try? decoder.decode(T.self, from: data)
    }
}
